app/src/main/java/org/mozilla/rocket/content/news/data/dailyhunt/DailyHuntNewsRemoteDataSource.kt

```diff
@@ -1,90 +1,177 @@
-import kotlinx.coroutines.Dispatchers
-import kotlinx.coroutines.withContext
+import android.net.Uri
+import androidx.paging.PageKeyedDataSource
 import mozilla.components.concept.fetch.MutableHeaders
 import mozilla.components.concept.fetch.Request
 import org.mozilla.focus.R
 import org.mozilla.rocket.content.Result
-import org.mozilla.rocket.content.news.data.NewsDataSource
+import org.mozilla.rocket.content.news.data.NewsDataSourceFactory.PageKey
 import org.mozilla.rocket.content.news.data.NewsItem
-import org.mozilla.rocket.util.safeApiCall
 import org.mozilla.rocket.util.sendHttpRequest
 import org.mozilla.rocket.util.sha256
 import org.mozilla.rocket.util.toJsonObject
 import java.net.URLEncoder
-import java.util.Locale
-
-class DailyHuntNewsRemoteDataSource(private val appContext: Context, private val newsProvider: DailyHuntProvider?) : NewsDataSource {
-
-    override suspend fun getNewsItems(category: String, language: String, pages: Int, pageSize: Int): Result<List<NewsItem>> = withContext(Dispatchers.IO) {
-        return@withContext safeApiCall(
-            call = {
-                val partner = newsProvider?.partnerCode ?: ""
-                val timestamp = System.currentTimeMillis().toString()
-                val uid = newsProvider?.userId ?: ""
-                sendHttpRequest(
-                    request = Request(
-                        url = getApiEndpoint(partner, timestamp, uid, category, language, pages, pageSize),
-                        method = Request.Method.GET,
-                        headers = createApiHeaders(partner, timestamp, uid, category, language, pages, pageSize)
-                    ),
-                    onSuccess = {
-                        Result.Success(fromJson(it.body.string()))
-                    },
-                    onError = {
-                        Result.Error(it)
-                    }
-                )
+
+class DailyHuntNewsRemoteDataSource(
+    private val appContext: Context,
+    private val newsProvider: DailyHuntProvider?,
+    private val category: String,
+    private val language: String
+) : PageKeyedDataSource<PageKey, NewsItem>() {
+
+    override fun loadInitial(params: LoadInitialParams<PageKey>, callback: LoadInitialCallback<PageKey, NewsItem>) {
+        val pageSize = params.requestedLoadSize
+
+        val result = fetchNewsItems(newsProvider, category, language,
+                pageSize, 0)
+        if (result is Result.Success) {
+            val (nextPageKey, items) = result.data
+            callback.onResult(items, null, nextPageKey)
+        } // TODO: error handling
+    }
+
+    override fun loadBefore(params: LoadParams<PageKey>, callback: LoadCallback<PageKey, NewsItem>) {
+        // Do nothing
+    }
+
+    override fun loadAfter(params: LoadParams<PageKey>, callback: LoadCallback<PageKey, NewsItem>) {
+        val pageKey = params.key as PageKey.PageUrlKey
+
+        val result = fetchNewsItemsNextPage(pageKey.url)
```

Comment:
I'm so curious if we should run this code in a CoroutineScope? We can inject the CoroutineScope to this class. Then we don't have to rely on the background executor from the paging library and we can cancel the job by ourselves.

Reply:
yeah, I created a ticket to mark this as a todo
https://github.com/mozilla-tw/FirefoxLite/issues/4928

Reply:
yeah, I created a ticket to mark this as a todo
#4928